src/Middleware/HeaderPropagation/src/HeaderPropagationMiddleware.cs

```diff
@@ -35,33 +35,43 @@ public HeaderPropagationMiddleware(RequestDelegate next, IOptions<HeaderPropagat
 
         public Task Invoke(HttpContext context)
         {
-            foreach ((var headerName, var entry) in _options.Headers)
+            // We need to intialize the headers because the message handler will use this to detect misconfiguration.
+            var headers = _values.Headers ??= new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
+
+            // Perf: avoid foreach since we don't define a struct enumerator.
+            var entries = _options.Headers;
+            for (var i = 0; i < entries.Count; i++)
             {
-                var values = GetValues(headerName, entry, context);
+                var entry = entries[i];
 
-                if (!StringValues.IsNullOrEmpty(values))
+                // We intentionally process entries in order, and allow earlier entries to
+                // take precedence over later entries when they have the same output name.
```

Comment:
Not sure what the convention is, but I would add a test to make sure this behavior stick as the comment can be missed.

Reply:
Sure 👍 Adding `MultipleEntries_AddsFirstToProduceValue`